futures-util/src/compat/compat01as03.rs

```diff
@@ -95,6 +138,119 @@ impl<St: Stream01> Stream03 for Compat01As03<St> {
+            inner: spawn01(inner),
+            buffer: None,
+        }
+    }
+
+    fn in_notify<R>(
+        &mut self,
+        lw: &LocalWaker,
+        f: impl FnOnce(&mut S) -> R,
+    ) -> R {
+        let notify = &WakerToHandle(lw.as_waker());
+        self.inner.poll_fn_notify(notify, 0, f)
+    }
+}
+
+impl<S, SinkItem> Stream03 for Compat01As03Sink<S, SinkItem>
+where
+    S: Stream01,
+{
+    type Item = Result<S::Item, S::Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Option<Self::Item>> {
+        match self.in_notify(lw, |f| f.poll()) {
+            Ok(Async01::Ready(Some(t))) => task03::Poll::Ready(Some(Ok(t))),
+            Ok(Async01::Ready(None)) => task03::Poll::Ready(None),
+            Ok(Async01::NotReady) => task03::Poll::Pending,
+            Err(e) => task03::Poll::Ready(Some(Err(e))),
+        }
+    }
+}
+
+impl<S, SinkItem> Sink03 for Compat01As03Sink<S, SinkItem>
+where
+    S: Sink01<SinkItem = SinkItem>,
+{
+    type SinkItem = SinkItem;
+    type SinkError = S::SinkError;
+
+    fn start_send(
+        mut self: Pin<&mut Self>,
+        item: Self::SinkItem,
+    ) -> Result<(), Self::SinkError> {
+        debug_assert!(self.buffer.is_none());
+        self.buffer = Some(item);
+        Ok(())
+    }
+
+    fn poll_ready(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Result<(), Self::SinkError>> {
+        match self.buffer.take() {
+            Some(item) => match self.in_notify(lw, |f| f.start_send(item)) {
+                Ok(AsyncSink01::Ready) => task03::Poll::Ready(Ok(())),
+                Ok(AsyncSink01::NotReady(i)) => {
+                    self.buffer = Some(i);
+                    task03::Poll::Pending
+                }
+                Err(e) => task03::Poll::Ready(Err(e)),
+            },
+            None => task03::Poll::Ready(Ok(())),
+        }
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Result<(), Self::SinkError>> {
+        let item = self.buffer.take();
+        match self.in_notify(lw, |f| match item {
+            Some(i) => match f.start_send(i) {
+                Ok(AsyncSink01::Ready) => f.poll_complete().map(|i| (i, None)),
+                Ok(AsyncSink01::NotReady(t)) => {
+                    Ok((Async01::NotReady, Some(t)))
+                }
+                Err(e) => Err(e),
+            },
+            None => Ok((Async01::NotReady, None)),
+        }) {
+            Ok((Async01::Ready(_), _)) => task03::Poll::Ready(Ok(())),
+            Ok((Async01::NotReady, item)) => {
+                self.buffer = item;
+                task03::Poll::Pending
+            }
+            Err(e) => task03::Poll::Ready(Err(e)),
+        }
+    }
+
+    fn poll_close(
+        self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Result<(), Self::SinkError>> {
+        self.poll_flush(lw)
```

Comment:
Not sure if this is the right way to `poll_close` this but from reading the docs all that is needed to close is just to flush and return `Poll::Ready`.

Reply:
There is a [`Sink01::close`](https://docs.rs/futures/0.1.25/futures/sink/trait.Sink.html#method.close) method that should be called here. I think you need to `poll_flush` _first_ then call the `close` method, 0.1 doesn't appear to make the same guarantee of flushing any remaining items when the sink is closed. Making this slightly more difficult is that once you call `Sink01::close` you can't call `poll_flush` anymore, so this probably needs an extra flag in the struct to know if you have started closing.

Reply:
Okay, I've attempted to fix this. `poll_complete` will only be called before close but never after, this is tracked via a bool field within the struct.